### PR TITLE
[API] get_balance

### PR DESCRIPTION
**Pull request by bernardoaraujor**
# Description of change

add `get_balance` API call
solve https://github.com/iotaledger/iota.c/issues/48

## How the change has been tested
[`tests/client/api_v1/test_get_balance.c`](https://github.com/bernardoaraujor/iota.c/blob/dev_get_balance/tests/client/api_v1/test_get_balance.c)

### CODE REVIEW

File: src/client/api/json_utils.c [hunk 1]
```diff
@@ -217,3 +217,189 @@ json_error_t json_get_uint64(cJSON const* const obj, char const key[], uint64_t*
 
   return JSON_OK;
 }
+
+int json_get_double(cJSON const* const json_obj, char const key[], double* const number) {
+  cJSON* json_value = cJSON_GetObjectItemCaseSensitive(json_obj, key);
+>>>>>>> Stashed changes
```
Comment (oopsmonk) [inline src/client/api/json_utils.c]: please fix merge conflicts in this file.

File: src/client/api/v1/get_balance.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+#include "core/utils/iota_str.h"
+
+int deser_balance_info(char const *const j_str, res_balance_t *res) {
+  char const *const key_addr = "address";
+  char const *const key_maxResults = "maxResults";
+  char const *const key_count = "count";
+  char const *const key_balance = "balance";
+  char const *const key_code = "code";
+  char const *const key_message = "message";
+  double *number;
```
Comment (oopsmonk) [inline src/client/api/v1/get_balance.c]: redundant variable

File: src/client/api/v1/get_balance.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+int get_balance(iota_client_conf_t const *conf, char *addr, res_balance_t *res) {
+  int ret = 0;
+  char const *const cmd_info = "api/v1/address/";
+
+  if (addr == NULL || res == NULL || conf == NULL) {
+    printf("[%s:%d]: get_balance failed (null parameter)\n", __func__, __LINE__);
+    return -1;
+  }
+
+  if (strlen(addr) != 64 && strlen(addr) != 81) {
```
Comment (oopsmonk) [inline src/client/api/v1/get_balance.c]: the legacy address is not supported.
```suggestion
  if (strlen(addr) != IOTA_ADDRESS_HEX_BYTES) {
```

File: src/client/api/v1/get_balance.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    res->balance = (int64_t)*number;
+
+  end:
+    cJSON_Delete(json_obj);
+    return ret;
+  }
+}
+
+int get_balance(iota_client_conf_t const *conf, char *addr, res_balance_t *res) {
```
Comment (oopsmonk) [inline src/client/api/v1/get_balance.c]: this parameter is an array not just a pointer, for readability I prefer
```suggestion
int get_balance(iota_client_conf_t const *conf, char addr[], res_balance_t *res) {
```

File: src/client/api/v1/get_balance.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  http_client_config_t http_conf = {0};
+  http_conf.url = cmd->buf;
+  if (conf->port) {
+    http_conf.port = conf->port;
+  }
+
+  byte_buf_t *http_res = byte_buf_new();
+  if (http_res == NULL) {
+    printf("[%s:%d]: OOM\n", __func__, __LINE__);
+    // TODO
```
Comment (oopsmonk) [inline src/client/api/v1/get_balance.c]: remove it

File: src/client/api/v1/get_balance.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#define __CLIENT_API_V1_BALANCE_H__
+
+#include <stdint.h>
+
+#include "client/client_service.h"
+#include "core/address.h"
+#include "core/types.h"
+
+typedef struct {
+  byte_t addr[IOTA_ADDRESS_BYTES];
```
Comment (oopsmonk) [inline src/client/api/v1/get_balance.h]: based on the implementation logic, it should be `IOTA_ADDRESS_HEX_BYTES`

File: tests/client/api_v1/test_get_balance.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  // test http status code 200
+  res_balance_t* res = calloc(1, sizeof(res_balance_t));
+  TEST_ASSERT_EQUAL_INT(0, deser_balance_info(json_info_200, res));
+  TEST_ASSERT_EQUAL_INT(1000, res->max_results);
+  TEST_ASSERT_EQUAL_INT(25, res->count);
+  TEST_ASSERT_EQUAL_INT(1338263, res->balance);
+
+  // test http status code 400
+  TEST_ASSERT_EQUAL_INT(-1, deser_balance_info(json_info_400, res));
```
Comment (oopsmonk) [inline tests/client/api_v1/test_get_balance.c:57]: please fix it as we discussed earlier: https://github.com/iotaledger/iota.c/pull/58#discussion_r537219009

File: tests/client/api_v1/test_get_balance.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      "\"code\": \"invalid_data\", "
+      "\"message\": \"invalid data provided\"}}";
+
+  char const* json_info_404 =
+      "{\"error\": {"
+      "\"code\": \"not_found\", "
+      "\"message\": \"could not find data\"}}";
+
+  // test http status code 200
+  res_balance_t* res = calloc(1, sizeof(res_balance_t));
```
Comment (oopsmonk) [inline tests/client/api_v1/test_get_balance.c]: dynamic allocation is not needed.
```suggestion
  res_balance_t res = {};
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]
thanks for your PR, please make sure compiling and testing are no problems in your local.

File: src/client/api/v1/get_balance.c (hunk 5 above)
Comment (bernardoaraujor) [reply src/client/api/v1/get_balance.c]: @oopsmonk you mean remove `// TODO`, or something else?

File: tests/client/api_v1/test_get_balance.c (hunk 7 above)
Comment (bernardoaraujor) [reply tests/client/api_v1/test_get_balance.c:57]: solved via ce64056c6d7d9453de0e78fda50f2f0a9410a080 and 2d2ce1682a05d2e8631d63a5d98856b318fca2da

File: tests/client/api_v1/test_get_balance.c (hunk 8 above)
Comment (bernardoaraujor) [reply tests/client/api_v1/test_get_balance.c]: solved via 9f583527292dd4bc9036f887dc0e75b7f02012e7

File: src/client/api/v1/get_balance.h (hunk 6 above)
Comment (bernardoaraujor) [reply src/client/api/v1/get_balance.h]: solved via ce64056c6d7d9453de0e78fda50f2f0a9410a080

**Comment by bernardoaraujor** [comment]
@oopsmonk is CI tests breaking because of memory leaks?
local tests are ok

**Comment by oopsmonk** [comment]
> @oopsmonk is CI tests breaking because of memory leaks?
> local tests are ok

hi @bernardoaraujor, you can see the CI report here: https://github.com/iotaledger/iota.c/runs/1533930414 or on [github actions](https://github.com/iotaledger/iota.c/actions).
it is a compiling issue on https://github.com/iotaledger/iota.c/pull/65/commits/47e3b00df9dcd9315e39167cfdeb37fc6ae7a456
 
could you make sure your local changes are as same as this PR? I believe you didn't change the `src/core/address.h` file but somehow it is changed in your commit, it might be caused by a conflicted merge.

**Comment by bernardoaraujor** [comment]
sorry @oopsmonk I think it's the `suggestions` feature from github.
I will stop using it since it's confusing me and reducing productivity.
I will open a new clean PR. closing this one.

**Pull request closed without merging**